Shipment Tracking: Provides option to delete a tracking record

anitaa1990 opened the pull request:
PR 4 of 5. 

Fixes #1037 . This PR adds logic to delete a shipment tracking from the Order Detail page and the Order Fulfillment page. 

### Items to be completed before PR can be submitted for review:
- [x] Order Detail page: remove icon in tracking item and add hamburger menu to delete/track shipment
- [x] Track shipment: current functionality if the user clicks the "open external" button. If no tracking link found, do not add this option to the menu list
- [x] Delete tracking - Clicking this option should call the FluxC action to delete the shipment tracking record and remove it from the list. Remove the shipment option immediately in the UI, then display the UNDO option to allow the user the chance to revert this action. If the UNDO snack is dismissed or times out, then the request to delete the shipment tracking record should be processed.
- [x] Order Fulfilment page: Add delete option UI to each tracking item
- [x] Delete tracking - Clicking this option should call the FluxC action to delete the shipment tracking record and remove it from the list. This should follow the same logic as the Order Detail page.
- [x] Events to be added for tracking purposes
- [x] Order detail page: hide the blank tracking card, if all items are deleted
- [x] Unit tests for the following scenarios:
     - [x] Order Detail page: Initiate delete request when no network connection. Expected Behaviour: Display offline snack and re add the deleted item to the tracking list
     - [x] Order Detail page: Initiate delete request when network is available.
         - [x] Success: Display success snack. 
         - [x] Error: Display error snack and re add the deleted item to the tracking list.
     - [x] Order Fulfil page: Initiate delete request when no network connection. Expected Behaviour: Display offline snack and re add the deleted item to the tracking list
     - [x]  Order Fulfil page: Initiate delete request when network is available.
         - [x] Success: Display success snack. 
         - [x] Error: Display error snack and re add the deleted item to the tracking list. 

### Notes
* [PR number 1](https://github.com/woocommerce/woocommerce-android/pull/1054) and [PR number 2](https://github.com/woocommerce/woocommerce-android/pull/1064) and [PR number 3](https://github.com/woocommerce/woocommerce-android/pull/1072) must be merged first.

### Screenshots

#### When there is only 1 tracking:
<img src="https://user-images.githubusercontent.com/22608780/58141716-f2483080-7c61-11e9-9825-d4f42333a87f.gif" width="300"/>

#### When there are multiple trackings:
<img src="https://user-images.githubusercontent.com/22608780/58078282-f7f33700-7bcb-11e9-8da9-a88946efd1cf.gif" width="300"/>

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

anitaa1990 commented:
@AmandaRiu @nbradbury merged the PR with the feature branch. Ready for review 👍

nbradbury commented:
When I tap the purple X, it doesn't act as though it was tapped. We could use a ripple or change of color when it's tapped.

![Screenshot_1558714476](https://user-images.githubusercontent.com/3903757/58342705-2026a400-7e1f-11e9-8bae-e89bcfffe59b.png)

nbradbury commented on RELEASE-NOTES.txt:
```diff
@@ -39,3 +39,4 @@ Improvements
 New Features
 * New support for the Shipment Tracking plugin. Read-only Shipment tracking information will now be displayed on a completed order and packages may be tracked directly from the order detail screen. Support for adding shipment tracking information is coming in a future release, so stay tuned!
```
Looks like we need to correct this.

anitaa1990 replied:
👍 Done in 6385e4d

AmandaRiu replied:
Added an additional comment to the changes.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailShipmentTrackingItemView.kt:
```diff
@@ -48,14 +53,52 @@ class OrderDetailShipmentTrackingItemView @JvmOverloads constructor(
+
+    /**
+     * Question: Should we count this action as non trivial action and
+     * call AppRatingDialog to incrementInteractions?
```
Good question! I'm fine with doing it here, and we should also do it after the user adds a tracking.

anitaa1990 replied:
Good point! I added this in 829f711, when:
* `Delete Tracking` is clicked in Order Detail page
* `Track Shipment` is clicked
* `Delete Tracking` icon is clicked in Order Fulfill page
* `Add` menu button is clicked in Add shipment tracking activity

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailShipmentTrackingListView.kt:
```diff
@@ -71,15 +72,32 @@ class OrderDetailShipmentTrackingListView @JvmOverloads constructor(
+
+    fun getShipmentTrackingCount() = shipmentTrack_items.adapter?.itemCount
+
     private fun enableItemAnimator(enable: Boolean) {
```
Looks like we're only calling this with `true`,  and we already setup the `itemAnimator` in `initView()`,  so we can just drop this method.

anitaa1990 replied:
Good point! Removed in 02c7878

nbradbury commented on WooCommerce/src/main/res/values/dimens.xml:
```diff
@@ -156,4 +156,10 @@
+    <!--
+        Shipment Trackings
+    -->
+    <dimen name="shipment_tracking_delete_button_width">30dp</dimen>
```
Are we sure we want the delete button to be 30dp? The minimum recommended tap target is 48dp.

anitaa1990 replied:
Setting the icon size bigger than 30dp was making it way too big. But you're right about the recommended target though. I increased the size to 32dp and wrapped the icon with a `FrameLayout` and gave extra padding to it so that the target target = 48dp. Done in a9c4865

anitaa1990 commented:
> When I tap the purple X, it doesn't act as though it was tapped. We could use a ripple or change of color when it's tapped.

I keep forgetting to add this! Thanks for pointing it out! Done in 5547760

Ready for round 2 👍

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailShipmentTrackingItemView.kt:
```diff
@@ -48,14 +54,51 @@ class OrderDetailShipmentTrackingItemView @JvmOverloads constructor(
             tracking_btnTrack.visibility = View.GONE
+            tracking_btnDelete.visibility = View.VISIBLE
+            tracking_btnDelete.setOnClickListener {
+                AnalyticsTracker.track(Stat.ORDER_TRACKING_DELETE, mapOf(
```
This is a data-related event and should be fired off when the request is submitted to delete the order shipment tracking option in the presenter. This is an important distinction because a lot can happen after the user clicks this button that would prevent us from _actually_ submitting the request to delete the object (such as network connectivity). We would only send this event when we are submitting the actual request to FluxC. Then the fail/success would also be handled in the presenter.

We want a view-related click event here... so similar to how we added a new event for clicking the "Add Tracking" button : `*_order_fulfillment_tracking_add_tracking_button_tapped`, we'll need a new track event for clicking the delete button. I recommend `*_order_fulfillment_tracking_delete_button_tapped` if on the order fulfillment page, and 
`*_order_detail_tracking_delete_button_tapped` if in order detail. Since this is a reusable component, it may make more sense to have this tracks event fired off by the fragment listener (for example inside the `listener?.deleteOrderShipmentTracking(item)`)

anitaa1990 replied:
Ahh 👌 I thought the `order_tracking_delete` event was view related event. Thanks for the clarification 👍 Modified in 0c2a4d0

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailShipmentTrackingItemView.kt:
```diff
@@ -48,14 +54,51 @@ class OrderDetailShipmentTrackingItemView @JvmOverloads constructor(
         }
+
+        popup.menu.findItem(R.id.menu_delete_shipment)?.setOnMenuItemClickListener {
+            AnalyticsTracker.track(Stat.ORDER_TRACKING_DELETE, mapOf(
```
Same comment as the one above regarding the use of this data-related track event.

anitaa1990 replied:
👍 Added events in 0c2a4d0

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailShipmentTrackingListView.kt:
```diff
@@ -64,22 +65,32 @@ class OrderDetailShipmentTrackingListView @JvmOverloads constructor(
 
-    private fun enableItemAnimator(enable: Boolean) {
-        shipmentTrack_items.itemAnimator = if (enable) DefaultItemAnimator() else null
+    fun deleteTrackingProvider(wcOrderShipmentTrackingModel: WCOrderShipmentTrackingModel) {
```
np: This naming is a bit misleading. A tracking provider would be the carrier IMO. I recommend renaming to `deleteTrackingRecord`

anitaa1990 replied:
Done in 98b3622

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailShipmentTrackingListView.kt:
```diff
@@ -64,22 +65,32 @@ class OrderDetailShipmentTrackingListView @JvmOverloads constructor(
+        shipmentTrack_items.swapAdapter(shipmentTrack_items.adapter, false)
     }
 
+    fun undoDeleteTrackingProvider(wcOrderShipmentTrackingModel: WCOrderShipmentTrackingModel) {
```
Same as above. Recommend renaming to `undoDeleteTrackingRecord`

anitaa1990 replied:
Makes sense! Done in 98b3622

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailShipmentTrackingListView.kt:
```diff
@@ -64,22 +65,32 @@ class OrderDetailShipmentTrackingListView @JvmOverloads constructor(
         private val trackings: MutableList<WCOrderShipmentTrackingModel>,
         private val uiMessageResolver: UIMessageResolver,
-        private val allowAddTrackingOption: Boolean
+        private val allowAddTrackingOption: Boolean,
```
Per the designs, the Add Tracking option is available on both the order detail and order fulfillment screens. I'm not sure if this is handled in a different PR, but thought I'd mention it here.

anitaa1990 replied:
I did not see the i8 version of the design spec and was working on the i7 version 🤦‍♀ . Started to implement this today but had a few doubts regarding how this will be displayed. For example, should the button be visible for all orders or just unfulfilled orders? Cesar has posted a [P2](https://hogsmeadep2.wordpress.com/2019/05/28/add-tracking-button-in-order-details/) on this. I will work on this after the discussion and create a separate PR?

AmandaRiu replied:
> I will work on this after the discussion and create a separate PR?

Works for me. I've opened a separate ticket to have this address: https://github.com/woocommerce/woocommerce-android/issues/1099

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderFulfillmentContract.kt:
```diff
@@ -28,5 +29,8 @@ interface OrderFulfillmentContract {
         fun showAddAddShipmentTrackingErrorSnack()
         fun toggleCompleteButton(isEnabled: Boolean)
         fun fulfillOrder()
+        fun reAddDeletedTrackingOnError()
```
np: recommend renaming to `undoDeletedTrackingOnError()`

anitaa1990 replied:
Makes sense! Done in 03a8cbc

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderFulfillmentFragment.kt:
```diff
@@ -185,6 +202,27 @@ class OrderFulfillmentFragment : Fragment(), OrderFulfillmentContract.View, View
     }
 
+    override fun reAddDeletedTrackingOnError() {
+        /**
```
np: Recommend moving the code comments to the method level.

anitaa1990 replied:
👍 Committed in 03a8cbc

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderFulfillmentFragment.kt:
```diff
@@ -228,4 +266,46 @@ class OrderFulfillmentFragment : Fragment(), OrderFulfillmentContract.View, View
     }
+
+    override fun deleteOrderShipmentTracking(item: WCOrderShipmentTrackingModel) {
+        /**
```
To avoid Kotlin Dokka confusion, I recommend using only `/*` instead of `/**` for inline code comments.

anitaa1990 replied:
Sure thing 👍 I have modified this in 03a8cbc

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderShipmentTrackingActionListener.kt:
```diff
@@ -1,5 +1,13 @@
+     * so adding a default implementation here to mark the method as optional so that
+     * implementing classes do not have to implement this method if not needed
+     */
+    fun openAddOrderShipmentTrackingScreen() { }
```
FYI, the ability to add tracking is available on both screens per the latest design specs.

anitaa1990 replied:
Ah, I did not see the i8 version of the design spec and was working on the i7 version 🤦‍♀ . Started to implement this today but had a few doubts regarding how this will be displayed. For example, should the button be visible for all orders or just unfulfilled orders? Cesar has posted a [P2](https://hogsmeadep2.wordpress.com/2019/05/28/add-tracking-button-in-order-details/) on this. I will work on this after the discussion and create a separate PR?

AmandaRiu replied:
Sounds good. opened a ticket to track: https://github.com/woocommerce/woocommerce-android/issues/1099

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailContract.kt:
```diff
@@ -50,5 +52,8 @@ interface OrderDetailContract {
         fun showLoadOrderError()
         fun refreshOrderStatus()
         fun refreshProductImages()
+        fun reAddDeletedTrackingOnError()
```
👍 Done in 03a8cbc

anitaa1990 commented on WooCommerce/src/main/res/layout/order_detail_shipment_tracking_item.xml:
```diff
@@ -47,25 +47,39 @@
-        android:layout_height="0dp"
-        android:background="?android:attr/selectableItemBackground"
+        style="@android:style/Widget.ActionButton"
+        android:layout_width="48dp"
```
👍 Done in b158f00

anitaa1990 commented on WooCommerce/src/main/res/layout/order_detail_shipment_tracking_item.xml:
```diff
@@ -47,25 +47,39 @@
-        android:clickable="true"
-        android:focusable="true"
-        app:layout_constraintBottom_toBottomOf="parent"
+        android:scaleType="center"
```
Nice catch! I did not notice this! Committed in b158f00

anitaa1990 commented on WooCommerce/src/main/res/layout/order_detail_shipment_tracking_item.xml:
```diff
@@ -47,25 +47,39 @@
+        app:layout_constraintTop_toTopOf="parent"
+        tools:visibility="visible"/>
+
+    <FrameLayout
```
I actually added it just for the padding but your idea of adding a `layer-list` is perfect 👌 I modified the button styles accordingly. Thank you 😊 Modified in b158f00

anitaa1990 commented on WooCommerce/src/main/res/drawable/delete_button_selector.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<ripple xmlns:android="http://schemas.android.com/apk/res/android"
```
👍 Modified this in b158f00

anitaa1990 commented on WooCommerce/src/main/res/values/styles.xml:
```diff
@@ -513,4 +513,12 @@
+    <!--
+        Delete Shipment Tracking
+    -->
+    <style name="Woo.Delete.Button" parent="android:TextAppearance.DeviceDefault">
```
👍 Done in b158f00

anitaa1990 commented on WooCommerce/src/main/res/values/dimens.xml:
```diff
@@ -156,4 +156,11 @@
+    -->
+    <dimen name="shipment_tracking_delete_button_width">32dp</dimen>
+    <dimen name="shipment_tracking_delete_button_height">32dp</dimen>
+    <dimen name="shipment_tracking_delete_button_padding">16dp</dimen>
```
👌 Done in b158f00

anitaa1990 commented:
@AmandaRiu thank you for the quick review 👍 I've modified the UI changes:
**1. Divider between the list items in shipment tracking:** Removing divider between the tracking records and it must only be visible if there is a tracking record available.

**2. The button looks significantly larger in the final version:** I actually added it just for the padding but your idea of adding a layer-list is perfect 👌 I modified the button styles accordingly. Thank you 😊

 Modified both of the above b158f00

Regarding adding tracking to Order Detail screen,  thank you pointing that out! I did not see the i8 version of the design spec and was working on the i7 version 🤦‍♀ . I started to implement this today but had a few doubts regarding how this will be displayed. For example, should the button be visible for all orders or just unfulfilled orders? Cesar has posted a [P2](https://hogsmeadep2.wordpress.com/2019/05/28/add-tracking-button-in-order-details/) on this. I will work on this after the discussion and create a separate PR.

Ready for another round of review 👍

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailShipmentTrackingListView.kt:
```diff
@@ -59,27 +59,53 @@ class OrderDetailShipmentTrackingListView @JvmOverloads constructor(
         }
     }
 
+    /*
```
To comply with [Kotlin KDoc guidelines](https://kotlinlang.org/docs/reference/kotlin-doc.html) and to ensure we will be able to use Kotlin Dokka in the future, please use the standard `\**` for method comments.

anitaa1990 replied:
Ah, ok. For inline comments I will use `//` and for method comments will use `/**`. I may have been doing this from the very beginning though so there might be a lot of places in the app where you might find this. I fixed the ones from this PR in b51f01f and will modify it as and when I see it in code. Sorry!

anitaa1990 commented on RELEASE-NOTES.txt:
```diff
@@ -37,5 +37,6 @@ Improvements
 
 New Features
-* New support for the Shipment Tracking plugin. Read-only Shipment tracking information will now be displayed on a completed order and packages may be tracked directly from the order detail screen. Support for adding shipment tracking information is coming in a future release, so stay tuned!
+* New support for the Shipment Tracking plugin. Read-only Shipment tracking information will now be displayed on a completed order and packages may be tracked directly from the order detail screen.
```
Good point! Since we would be adding the `Add Tracking` button to Order Detail page in a separate PR, would it be ok to add this change to that PR? Technically, if we run changes in this PR, the add tracking functionality would only be run for unfulfilled orders. So it makes more sense to modify this when fixing #1099

anitaa1990 commented:
> I tested the app and it's crashing when I attempt to track a package from the Order Detail view. 

I believe this was fixed in this [PR](https://github.com/woocommerce/woocommerce-android/pull/1093) and merged to `develop` branch and the `feature/add-shipment-tracking` branch. I don't think I merged it with this one though till you mentioned it. So I went ahead and did in d0a59b2

> I also discovered that if you delete two shipment tracking records back-to-back, the confirmation snackbar for deleting the first record will dismiss the UNDO snackbar for the second request. So if the user deletes two tracking records but only meant to delete the first they cannot undo

This is a little tricker. I added logic to dismiss the first snackbar if a second item from the list is deleted in b51f01f. But I forsee some scenarios I might not have tested so I will add some extra UI tests on this to ensure that every scenario is taking care of, in this [PR](https://github.com/woocommerce/woocommerce-android/pull/1094).

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailFragment.kt:
```diff
@@ -426,12 +441,84 @@ class OrderDetailFragment : Fragment(), OrderDetailContract.View, OrderDetailNot
     }
 
+    override fun deleteOrderShipmentTracking(item: WCOrderShipmentTrackingModel) {
+        /**
```
Done in b51f01f. As I mentioned in the below comment, I may have been doing this from the very beginning though so there might be a lot of places in the app where you might find this. I fixed the ones from this PR and will modify it as and when I see it in code. Sorry!

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailFragment.kt:
```diff
@@ -426,12 +441,84 @@ class OrderDetailFragment : Fragment(), OrderDetailContract.View, OrderDetailNot
     }
 
+    override fun undoDeletedTrackingOnError() {
+        /**
```
Done in b51f01f

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailFragment.kt:
```diff
@@ -426,12 +440,93 @@ class OrderDetailFragment : Fragment(), OrderDetailContract.View, OrderDetailNot
+            deleteOrderShipmentTrackingSnackbar = null
+        }
+
+        AnalyticsTracker.track(ORDER_DETAIL_TRACKING_DELETE_BUTTON_TAPPED)
```
👍 Done in d591042
